README.md
# json_ingest_chatbot

<p align="center"><img src="https://socialify.git.ci/tomkat-cr/json_ingest_chatbot/image?description=0&amp;font=Inter&amp;language=1&amp;name=1&amp;owner=1&amp;pattern=Plus&amp;stargazers=0&amp;theme=Light" alt="project-image"></p>

JSON ingest chatbot using Python, Langchain and OpenAI GPT models

This repository contains the source code for a chatbot application that interacts with multiple JSON data documents and/or Git repositories. The application uses `OpenAI` GPT models to generate conversational responses based on the contents of the JSON files, Git repositories and other sources, and `streamlit` for the web interface.

<!---
![Chat with Multiple JSONs](assets/screen.png)
-->

## Repository Contents

- `src/main.py`: This script runs the main part of the application. It includes functionalities such as JSON files uploading, input of Git repository URL or local path, extracting data from JSON files, handling user inputs, generating conversational responses, and setting up the Streamlit application.
- `vector_index.py`: all vector related operations, using Langchain's VectorstoreIndexCreator and other conversational Langchain libraries.
- `src/json_reader.py`: JSON files loader (all loaders use Langchain document loader libraries).
- `src/git_reader.py`: Git repository URL or local path loader.
- `src/pdf_reader.py`: PDF files loader.
- `src/youtube_reader.py`: Youtube video loader.
- `src/htmlcss.py`: Contains the HTML and CSS templates used to structure and style the web application.

## How to Run the Application

To run the application, execute the following command in the terminal:

```bash
streamlit run src/json.py
```

You will also need to provide your own API key to access OpenAI.

## Instructions for Use

1. Run the application with the command above.

2. Navigate to the browser window that opens. Usually it's: http://localhost:8501/

3. In the sidebar, upload the JSON files you want the chatbot to interact with.

4. Configure the GPT model and temperature options as desired.

5. Click "Process" and wait for the JSON files to be processed.

6. Enter your questions in the input box and receive responses from the chatbot.

## Question examples:

Write a readme.md file for this repository content.
Give me a pytest file for this repository content.
Describe the supplied context files.
Describe the supplied context.

## Contributors

[Your Name] - [Your Email]
[Contributor Name] - [Contributor Email]

Please feel free to suggest improvements, report bugs, or make a contribution to the code.

## License

This project is licensed under the terms of the MIT license.

## Acknowledgements

This project uses OpenAI's GPT models and Streamlit. We appreciate their contributions to the open-source community.
Readme generate by [The AI Readme Generator](https://github.com/tomkat-cr/ai_readme_generator/)

CHANGELOG.md
# CHANGELOG

All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/) and [Keep a Changelog](http://keepachangelog.com/).



## Unreleased
---

### New

### Changes

### Fixes

### Breaks


## 0.1.2 (2023-07-21)
---

### New
Add PDF and Youtube loaders (only the loader for now).

### Changes
Separation of concerns: `main.py` replaces `json.py` as the main script, `vector_index.py` has all vector related operations, and each loader is separated into its own code file: `json_reader.py`, `git_reader.py`, `pdf_reader.py`, and `youtube_reader.py`.


## 0.1.1 (2023-07-20)
---

### New
Add Git repository reading.
Add visual design as ChatGPT, with the input text at the bottom.

### Fixes
JSON ingest assigns page_content with the whole json content.
Fix broken icon images. The only way to do it to use a public URL in the `src` attribute. Locals does not work.
JSON file `page_content` is not the file title, is the file content as serialized string. This expands the context.


## 0.1.0 (2023-06-29)
---

### New
First JSON ingest version with fixing assignment of page_content based on the JSON filename.


src/main.py
from dotenv import load_dotenv

import streamlit as st
from htmlcss import bot_template, user_template, css

from json_reader import get_json_files
from git_reader import get_repo_data
from vector_index import get_vector_index, get_conversation_chain


def handle_userinput(user_question):
    if st.session_state.conversation is None:
        return

    response = st.session_state.conversation({'question': user_question})
    st.session_state.chat_history = response['chat_history']

    for i, message in enumerate(st.session_state.chat_history):
        if i % 2 == 0:
            st.write(user_template.replace(
                "{{MSG}}", message.content), unsafe_allow_html=True)
        else:
            st.write(bot_template.replace(
                "{{MSG}}", message.content), unsafe_allow_html=True)


def main():
    load_dotenv()

    st.set_page_config(
        page_title="Chat with JSON files and Git repos",
        page_icon=":card_index_dividers:"
    )
    st.write(css, unsafe_allow_html=True)

    if "conversation" not in st.session_state:
        st.session_state.conversation = None
    if "chat_history" not in st.session_state:
        st.session_state.chat_history = None

    st.header(
        "Chat with JSON files and Git repos"
    )

    # user_question = st.text_input(
    user_question = st.chat_input(
        "Ask something to the JSON or Git files:"
    )
    if user_question:
        handle_userinput(user_question)

    with st.sidebar:
        st.subheader("JSON files")
        json_docs = st.file_uploader(
            "Upload the JSON file(s) here and click on 'Process'",
            accept_multiple_files=True
        )

        st.subheader("Git repo URL")
        repo_url = st.text_input(
            "Enter the Git repo URL here and click on 'Process'",
            value="",
            # max_chars=80,
            placeholder="https://github.com/username/repo.git"
        )
        branch = st.text_input(
            "Git brach [main]:",
            value="",
            # max_chars=80,
            placeholder="main"
        )

        st.subheader("Model Configuration")
        gpt_model = st.selectbox(
            'Model',
            ('gpt-3.5-turbo', 'gpt-4')
        )
        gpt_temperature = st.slider('Temperature', 0.00, 1.00, 0.7)

        if st.button("Process"):
            with st.spinner("Procesing..."):
                if branch == "":
                    branch = "main"
                repo_data = get_repo_data(repo_url, branch)
                all_json_files = get_json_files(json_docs)
                # index = VectorstoreIndexCreator().from_documents(
                #     all_json_files + repo_data
                # )
                index = get_vector_index(
                    all_json_files + repo_data
                )
                st.session_state.conversation = get_conversation_chain(
                    index.vectorstore,
                    gpt_model,
                    gpt_temperature
                )

        st.markdown('Click on **Process** to load the ' +
                    'JSON files and/or the Git repo.')


if __name__ == '__main__':
    main()


src/htmlcss.py
img_width = "28px"
img_height = "28px"
img_basse_path = "https://github.com/tomkat-cr/json_ingest_chatbot/raw/main"
icon_chatbot = f"{img_basse_path}/assets/icon_chatbot.jpeg"
icon_human = f"{img_basse_path}/assets/icon_human.jpeg"

css = '''
<style>
.chat-message {
    padding: 1.5rem; border-radius: 0.5rem; margin-bottom: 1rem; display: flex
}
.chat-message.user {
    background-color: #2b313e
}
.chat-message.bot {
    background-color: #475063
}
.chat-message .avatar {
  width: 5%;
}
.chat-message .avatar img {
  max-width: ''' + img_width + ''';
  max-height: ''' + img_height + ''';
  border-radius: 50%;
  object-fit: cover;
}
.chat-message .message {
  width: 95%;
  padding: 0 0.5rem;
  color: #fff;
}
'''

bot_template = '''
<div class="chat-message bot">
    <div class="avatar">
        <img
            src="''' + icon_chatbot + '''"
            style="max-height: ''' + img_height + ''';
            max-width: ''' + img_width + ''';''' + \
            ''' border-radius: 50%; object-fit: cover;"
        >
    </div>
    <div class="message">{{MSG}}</div>
</div>
'''

user_template = '''
<div class="chat-message user">
    <div class="avatar">
        <img
            src="''' + icon_human + '''"
            style="max-height: ''' + img_height + ''';
            max-width: ''' + img_width + ''';''' + \
            ''' border-radius: 50%; object-fit: cover;"
        >
    </div>
    <div class="message">{{MSG}}</div>
</div>
'''


src/git_reader.py
import os
from pprint import pprint

from git import Repo
from langchain.document_loaders import GitLoader

from commons import DEBUG, VectorEntry, std_response


def remove_dir(local_temp_path):
    if local_temp_path == '/':
        raise Exception("Cannot use / as a temp path")
    if not os.path.exists(local_temp_path):
        return
    if os.path.isfile(local_temp_path):
        os.remove(local_temp_path)
        return
    for root, dirs, files in os.walk(local_temp_path, topdown=False):
        for name in files:
            os.remove(os.path.join(root, name))
        for name in dirs:
            os.rmdir(os.path.join(root, name))


def get_repo(repo_url, branch):
    response = std_response()
    repo_name = repo_url.rsplit('/', 1)[1]
    if repo_url[:8] == "https://":
        repo_type = "remote"
        # https://python.langchain.com/docs/modules/data_connection/document_loaders/integrations/git
        local_temp_repo_path = f"/tmp/{repo_name}"
        try:
            remove_dir(local_temp_repo_path)
        except Exception as err:
            if DEBUG:
                raise
            response["error"] = True
            response["error_msg"] = f"ERROR: {str(err)}"
            return response
        try:
            repo = Repo.clone_from(
                url=repo_url,
                to_path=local_temp_repo_path
            )
        except Exception as err:
            if DEBUG:
                raise
            response["error"] = True
            response["error_msg"] = f"ERROR: {str(err)}"
            return response
    else:
        repo_type = "local"
        try:
            local_temp_repo_path = repo_url
            repo = Repo(local_temp_repo_path)
        except Exception as err:
            if DEBUG:
                raise
            response["error"] = True
            response["error_msg"] = f"ERROR: {str(err)}"
            return response
    if not branch:
        try:
            branch = repo.head.reference
        except Exception as err:
            if DEBUG:
                raise
            response["error"] = True
            response["error_msg"] = f"ERROR: {str(err)}"
            return response
    print()
    print(f"Github repository URL: {repo_url}")
    print(f"Branch (default 'main'): {branch}")
    try:
        loader = GitLoader(
            repo_path=f"{local_temp_repo_path}/",
            branch=branch,
        )
    except Exception as err:
        if DEBUG:
            raise
        response["error"] = True
        response["error_msg"] = f"ERROR: {str(err)}"
        return response

    response["data"] = loader.load()

    # Add some context for the repo
    metadata = {
        "source": repo_url,
        "branch": branch,
        "repo_type": repo_type,
        "repo_name": repo_name,
        "repo_path": local_temp_repo_path,
        "comments": ",".join([
                    "this is the context of the:",
                    "repo context",
                    "repository context",
                    "supplied repo"
                    "supplied repository",
                    "loaded repo",
                    "loaded repository",
                    "supplied git repo",
                    "supplied git repository",
                    "loaded git repo",
                    "loaded git repository",
                    "supplied branch",
                    "loaded branch",
            ])
    }
    response["data"].append(
        VectorEntry(
            page_content=f"The supplied git repo name is {repo_name}",
            metadata=metadata
        )
    )
    response["data"].append(
        VectorEntry(
            page_content=f"{repo_type} git repo content of {repo_url}",
            metadata=metadata
        )
    )

    if DEBUG:
        print(f"Repo: {repo_url}")
        print(f"Branch: {branch}")
        print("Data:")
        pprint(response["data"])
    return response


def get_repo_data(repo_url, branch):
    if repo_url == "":
        return []
    repo_response = get_repo(repo_url, branch)
    if repo_response["error"]:
        return []
    repo_data = repo_response["data"]
    return repo_data


src/json_reader.py
import json
from langchain.document_loaders.json_loader import JSONLoader

from commons import DEBUG, VectorEntry


def get_json_files(json_files):
    all_pages = []
    for json_file in json_files:
        if DEBUG:
            print(f"json_file: {json_file}")
        output_file_spec = f"./downloads/{json_file.name}"
        if not json_file.name.endswith(".json"):
            continue
        with open(output_file_spec, 'wb') as output_file:
            json_content = json_file.getvalue()
            # json_dict = {
            #     'content': json.loads(json_content)
            # }
            # json_dict[attr_name] = get_page_content_attr(json_file.name)
            json_dict = {}
            json_dict['content'] = json.loads(json_content)
            # The attribute 'page_content' must be a string,
            # so the json content must be serialized
            page_content = json.dumps(json_dict['content'])
            json_dict['page_content'] = page_content
            json_content = bytes(json.dumps(json_dict), 'UTF8')
            if DEBUG:
                print(f"json_content: {json_content}")
            output_file.write(json_content)
        loader = JSONLoader(
            output_file_spec,
            # jq_schema=f'.[].{attr_name}'
            jq_schema='.page_content'
        )

        json_pages = loader.load()

        # Add some context for the JSON file
        metadata = {
            'source': output_file_spec,
            'file_path': output_file_spec,
            'file_name': json_file.name,
            'file_type': '.json',
            "comments": ",".join([
                "this is the context and content of the:",
                "supplied json file",
                "loaded json file"
                "supplied json",
                "loaded json",
            ])
        }
        json_pages.append(
            VectorEntry(
                page_content=f"The loaded JSON filename is {json_file.name}",
                metadata=metadata
            )
        )

        all_pages += json_pages
        # all_pages.append(json_pages)

    return all_pages


src/vector_index.py
from langchain.chat_models import ChatOpenAI
from langchain.memory import ConversationBufferMemory
from langchain.chains import ConversationalRetrievalChain
from langchain.indexes import VectorstoreIndexCreator


class VectorEntry():
    def __init__(self, page_content, metadata={}):
        # Example: page_content='The loaded git repo name is myrepo'
        self.page_content = page_content
        #  Example: metadata={'source': 'path/to/file.py',
        # 'file_path': 'path/to/file.py', 'file_name': 'file.py',
        # 'file_type': '.py'}
        self.metadata = metadata


def get_vector_index(vectorstore):
    index = VectorstoreIndexCreator().from_documents(
        vectorstore
    )
    return index


def get_conversation_chain(
    vectorstore, gpt_model, gpt_temperature
):
    llm = ChatOpenAI(
        model=gpt_model,
        temperature=gpt_temperature
    )
    memory = ConversationBufferMemory(
        memory_key='chat_history', return_messages=True
    )
    conversation_chain = ConversationalRetrievalChain.from_llm(
        llm=llm,
        retriever=vectorstore.as_retriever(),
        memory=memory
    )
    return conversation_chain


scripts/install_dev_tools.sh
#!/bin/bash
ERROR_MSG=""

if [ "${ERROR_MSG}" = "" ]; then
    export SCRIPTS_DIR="$(dirname "$0")";
    cd "${SCRIPTS_DIR}"
    export SCRIPTS_DIR="$(pwd)";
    cd "${SCRIPTS_DIR}/.."
    export BASE_DIR="$(pwd)";
    if [ -f "${BASE_DIR}/.env" ]; then
        set -o allexport; . "${BASE_DIR}/.env" ; set +o allexport ;
    fi
fi

if [ "${PYTHON_VERSION}" == "" ]; then
    PYTHON_VERSION="3.9.17"
fi

BASH_PROFILE="${HOME}/.bash_profile"
BASH_SOURCE="bash"
if [ -f "${HOME}/.zshrc" ]; then
    BASH_PROFILE="${HOME}/.zshrc"
    BASH_SOURCE="zsh"
fi

PYENV_INSTALL_PREFIX=""
MACHINE_ARCHITECTURE=$(arch)
if [ "${MACHINE_ARCHITECTURE}" = "arm64" ]; then
    PYENV_INSTALL_PREFIX="arch -x86_64"
fi

if [ "${ERROR_MSG}" = "" ]; then
    if ! pyenv --version
    then
        # install pyenv
        echo ""
        echo "Install pyenv"
        echo ""
        curl https://pyenv.run | bash
        if ! pyenv --version
        then
            ERROR_MSG="ERROR: pyenv could not be installed."
        fi
    fi
fi

if [ "${ERROR_MSG}" = "" ]; then
    if ! grep -q 'eval "$(pyenv init --path)"' ${BASH_PROFILE}; then
        # add pyenv to the PATH and enable shims and autocompletion
        echo ""
        echo "Add pyenv to the PATH and enable shims and autocompletion"
        echo ""

        echo "" >> ${BASH_PROFILE}
        echo "# `date`" >> ${BASH_PROFILE}
        echo "# Add pyenv to the PATH and enable shims and autocompletion" >> ${BASH_PROFILE}
        echo "" >> ${BASH_PROFILE}
        echo 'export PYENV_ROOT="${HOME}/.pyenv"' >> ${BASH_PROFILE}
        echo 'export PATH="$PYENV_ROOT/bin:$PATH"' >> ${BASH_PROFILE}
        echo 'eval "$(pyenv init --path)"' >> ${BASH_PROFILE}
        echo 'eval "$(pyenv init -)"' >> ${BASH_PROFILE}
        # echo 'eval "$(pyenv virtualenv-init -)"' >> ${BASH_PROFILE}
    fi
fi

if [ "${ERROR_MSG}" = "" ]; then
    if [ ! -d "${HOME}/.pyenv/versions/${PYTHON_VERSION}" ]; then
        # install python version using pyenv
        echo ""
        echo "Install python with: pyenv install ${PYTHON_VERSION}"
        echo ""
        if ! ${PYENV_INSTALL_PREFIX} pyenv install ${PYTHON_VERSION}
        then
            ERROR_MSG="ERROR: running pyenv install ${PYTHON_VERSION}"
        fi
    fi
fi

if [ "${ERROR_MSG}" = "" ]; then
    echo ""
    echo "Changing python version: pyenv local ${PYTHON_VERSION}"
    echo ""
    if ! pyenv local ${PYTHON_VERSION}
    then
        ERROR_MSG="ERROR: running pyenv local ${PYTHON_VERSION}"
    fi
fi

if [ "${ERROR_MSG}" = "" ]; then
    if ! pipenv --version
    then
        # install pipenv
        echo ""
        echo "Install pipenv"
        echo ""
        if ! pip install --user pipenv
        then
            ERROR_MSG="ERROR: pipenv could not be installed."
        else
            # add pipenv to the bash profile
            ADD_CMD_TO_PROFILE="eval \"$(_PIPENV_COMPLETE=${BASH_SOURCE}_source pipenv)\""
            if ! grep -q "${ADD_CMD_TO_PROFILE}" ${BASH_PROFILE}; then
                echo ""
                echo "Add pipenv to the ${BASH_SOURCE} profile"
                echo ""

                echo "" >> ${BASH_PROFILE}
                echo "# `date`" >> ${BASH_PROFILE}
                echo "# Add pipenv init" >> ${BASH_PROFILE}
                echo "" >> ${BASH_PROFILE}
                echo "${ADD_CMD_TO_PROFILE}" >> ${BASH_PROFILE}
            fi
        fi
    fi
fi

if [ "${ERROR_MSG}" = "" ]; then
    if ! make --version
    then
        # install make
        echo ""
        echo "Install make"
        echo ""

        if [[ "$OSTYPE" == "darwin"* ]]; then
            brew install make
        else
            if ! apt -y install make
            then
                if ! yum -y install make
                then
                    if ! apk add make
                    then
                        ERROR_MSG="ERROR: make could not be installed in linux (debian/rhel/alpine)"
                    fi
                fi
            fi  
        fi
        if ! make --version
        then
            ERROR_MSG="ERROR: make could not be installed"
        fi
    fi
fi

if [ "${ERROR_MSG}" = "" ]; then
    if ! aws --version
    then
        sh ${SCRIPTS_DIR}/install_aws_cli.sh
        if ! aws --version
        then
            ERROR_MSG="ERROR: aws-cli could not be installed"
        fi
    fi
fi

if [ "${ERROR_MSG}" = "" ]; then
    echo ""
    echo "Reload the shell"
    echo ""
    exec "$SHELL"
fi

echo ""
if [ "${ERROR_MSG}" = "" ]; then
    echo "Done!"
else
    echo "${ERROR_MSG}" ;
fi
echo ""


src/json.py
import json
import streamlit as st
from dotenv import load_dotenv
from htmlcss import bot_template, user_template, css
from langchain.document_loaders.json_loader import JSONLoader
from langchain.chat_models import ChatOpenAI
from langchain.memory import ConversationBufferMemory
from langchain.chains import ConversationalRetrievalChain
from langchain.indexes import VectorstoreIndexCreator


def get_page_content_attr(filename):
    attr_value = filename
    if filename == "general_ingredients.json":
        attr_value = "general ingredients table"
    elif filename == "food_times.json":
        attr_value = "meal type"
    elif filename == "user_ingredients.json":
        attr_value = "my favorite ingredients"
    elif filename == "food_history.json":
        attr_value = "feeding history, dietary preferences"
    elif filename == "user.json":
        attr_value = "my personal information, dietary restrictions, " + \
            "and any specific dietary goals"
    return attr_value


def get_json_files(json_files):
    attr_name = 'page_content'
    all_pages = []
    for json_file in json_files:
        print(f"json_file: {json_file}")
        output_file_spec = f"./downloads/{json_file.name}"
        with open(output_file_spec, 'wb') as output_file:
            json_content = json_file.getvalue()
            json_dict = {
                'content': json.loads(json_content)
            }
            json_dict[attr_name] = get_page_content_attr(json_file.name)
            json_content = bytes(json.dumps(json_dict), 'UTF8')
            print(f"json_content: {json_content}")
            output_file.write(json_content)
        loader = JSONLoader(
            output_file_spec,
            # jq_schema=f'.[].{attr_name}'
            jq_schema=f'.{attr_name}'
        )
        json_pages = loader.load()
        all_pages += json_pages
    return all_pages


def handle_userinput(user_question):
    response = st.session_state.conversation({'question': user_question})
    st.session_state.chat_history = response['chat_history']

    for i, message in enumerate(st.session_state.chat_history):
        if i % 2 == 0:
            st.write(user_template.replace(
                "{{MSG}}", message.content), unsafe_allow_html=True)
        else:
            st.write(bot_template.replace(
                "{{MSG}}", message.content), unsafe_allow_html=True)


def get_conversation_chain(
    vectorstore, gpt_model, gpt_temperature
):
    llm = ChatOpenAI(
        model=gpt_model,
        temperature=gpt_temperature
    )
    memory = ConversationBufferMemory(
        memory_key='chat_history', return_messages=True
    )
    conversation_chain = ConversationalRetrievalChain.from_llm(
        llm=llm,
        retriever=vectorstore.as_retriever(),
        memory=memory
    )
    return conversation_chain


def main():
    load_dotenv()

    st.set_page_config(
        page_title="Chat con multiples JSONs",
        page_icon=":books:"
    )
    st.write(css, unsafe_allow_html=True)

    if "conversation" not in st.session_state:
        st.session_state.conversation = None
    if "chat_history" not in st.session_state:
        st.session_state.chat_history = None

    st.header(
        "Chat con multiples JSONs :books:"
    )

    user_question = st.text_input(
        "Pregúntale lo que quieras a tus documentos JSON:"
    )
    if user_question:
        handle_userinput(user_question)

    with st.sidebar:

        st.subheader("Configuración del Modelo")
        gpt_model = st.selectbox(
            'Modelo',
            ('gpt-3.5-turbo', 'gpt-4')
        )
        gpt_temperature = st.slider('Temperatura', 0.00, 1.00, 0.7)

        st.subheader("Tus archivos")
        json_docs = st.file_uploader(
            "Sube tus Json aquí y haz click en 'Procesar'",
            accept_multiple_files=True
        )
        if st.button("Procesar"):
            with st.spinner("Procesando..."):
                all_json_files = get_json_files(json_docs)
                index = VectorstoreIndexCreator().from_documents(
                    all_json_files
                )
                st.session_state.conversation = get_conversation_chain(
                    index.vectorstore,
                    gpt_model,
                    gpt_temperature
                )


if __name__ == '__main__':
    main()
